File: c++/c20_4_path_covers.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
constexpr ll INF = numeric_limits<ll>::max() / 2;
#include "c13_3_floyd_warshall_algorithm.h"
#include "c20_3_maximum_matchings.h"

namespace {
	struct Connect {
		Connect(ll l) : adj(l), start(views::iota(1ll, l) | ranges::to<unordered_set>()) {}
		void add(ll i, ll j) {
			adj[i].push_back(j);
			start.erase(j);
		}
		vector<vector<ll>> routes() {
			vector<vector<ll>> ret;
			for (ll i : start) {
				ret.resize(ret.size() + 1);
				auto& r = ret.back();
				r.push_back(i);
				while (!adj[r.back()].empty()) {
					ll j = r.back();
					r.push_back(adj[j].back());
					adj[j].pop_back();
				}
			}
			return ret;
		}
	private:
		vector<vector<ll>> adj;
		unordered_set<ll> start;
	};

	vector<vector<ll>> node_disjoint(const vector<vector<ll>>& adj) {
		ll l = ssize(adj);
		vector<pair<ll, ll>> edges;
		for (ll i = 1; i < l; ++i) {
			for (ll j : adj[i]) {
				edges.emplace_back(i, j + l);
			}
		}
		Connect con(l);
		for (auto [i, j] : maximum_matching(edges))
			con.add(i, j - l);
		return con.routes();
	}

	vector<vector<ll>> general(const vector<vector<ll>>& adj) {
		ll l = ssize(adj);
		vector<vector<ll>> adj2(l, vector<ll>(l));
		for (ll i = 1; i < l; ++i) {
			for (ll j : adj[i]) {
				adj2[i][j] = 1;
			}
		}
		auto distance = floyd_warshall(adj2);
		vector<pair<ll, ll>> edges;
		for (ll i = 1; i < l; ++i) {
			for (ll j = 1; j < l; ++j) {
				if (0 < distance[i][j] && distance[i][j] < INF) {
					edges.emplace_back(i, j + l);
				}
			}
		}
		Connect con(l);
		for (auto [i, j] : maximum_matching(edges)) {
			j -= l;
			while (distance[i][j] > 1) {
				for (ll k = 1; k < l; ++k) {
					if (distance[k][j] == distance[i][j] - 1) {
						con.add(i, k);
						i = k;
						break;
					}
				}
			}
			con.add(i, j);
		}
		return con.routes();
	}

	void test_node_disjoint(const vector<vector<ll>>& adj, ll l) {
		auto ret = node_disjoint(adj);
		EXPECT_EQ(ret.size(), l);
		unordered_multiset<ll> all;
		for (const auto& r : ret)
			all.insert(r.begin(), r.end());
		EXPECT_EQ(all, views::iota(1ll, ll(ssize(adj))) | ranges::to<unordered_multiset>());
		for (const auto& r : ret) {
			for (auto [a, b] : r | views::pairwise) {
				EXPECT_TRUE(ranges::find(adj[a], b) != adj[a].end());
			}
		}
	}

	void test_general(const vector<vector<ll>>& adj, ll l) {
		auto ret = general(adj);
		EXPECT_EQ(ret.size(), l);
		unordered_set<ll> all;
		for (const auto& r : ret)
			all.insert(r.begin(), r.end());
		EXPECT_EQ(all, views::iota(1ll, ll(ssize(adj))) | ranges::to<unordered_set>());
		for (const auto& r : ret) {
			for (auto [a, b] : r | views::pairwise) {
				EXPECT_TRUE(ranges::find(adj[a], b) != adj[a].end());
			}
		}
	}
}

TEST(C204PathCovers, nodeDisjoint) {
	test_node_disjoint({ {},{2},{} }, 1);
	test_node_disjoint({ {},{5},{6},{4},{},{6},{3,7},{} }, 3);
}

TEST(C204PathCovers, general) {
	test_general({ {},{2},{} }, 1);
	test_general({ {},{5},{6},{4},{},{6},{3,7},{} }, 2);
}

File: python/c20_4_path_covers.py
import unittest
from itertools import pairwise
from c13_3_floyd_warshall_algorithm import INF,floyd_warshall
from c20_3_maximum_matchings import maximum_matching

class Connect:
    def __init__(self,l):
        self.adj=[[] for _ in range(l)]
        self.start={*range(1,l)}
    def add(self,i,j):
        self.adj[i].append(j)
        self.start.discard(j)
    def routes(self):
        for i in self.start:
            r=[i]
            while self.adj[r[-1]]:
                r.append(self.adj[r[-1]].pop())
            yield r

def node_disjoint(adj):
    l=len(adj)
    con=Connect(l)
    for i,j in maximum_matching([(i,j+l) for i in range(1,len(adj)) for j in adj[i]]):
        con.add(i,j-l)
    return con.routes()

def general(adj):
    l=len(adj)
    adj2=[[0]*l for _ in range(l)]
    for i in range(1,len(adj)):
        for j in adj[i]:
            adj2[i][j]=1
    distance=floyd_warshall(adj2)
    con=Connect(l)
    for i,j in maximum_matching([(i,j+l) for i in range(1,l) for j in range(1,l) if 0<distance[i][j]<INF]):
        j-=l
        while distance[i][j]>1:
            k=next(k for k in range(1,l) if distance[k][j]==distance[i][j]-1)
            con.add(i,k)
            i=k
        con.add(i,j)
    return con.routes()

class Test(unittest.TestCase):
    def assertNodeDisjoint(self,adj,l):
        with self.subTest(adj=adj):
            ret=[*node_disjoint(adj)]
            self.assertEqual(len(ret),l)
            self.assertCountEqual((a for r in ret for a in r), range(1,len(adj)))
            for r in ret:
                for a,b in pairwise(r):
                    self.assertIn(b,adj[a])
    def assertGeneral(self,adj,l):
        with self.subTest(adj=adj):
            ret=[*general(adj)]
            self.assertEqual(len(ret),l)
            self.assertCountEqual(set(a for r in ret for a in r), range(1,len(adj)))
            for r in ret:
                for a,b in pairwise(r):
                    self.assertIn(b,adj[a])
    def test(self):
        adj=[[],[2],[]]
        self.assertNodeDisjoint(adj,1)
        self.assertGeneral(adj,1)

        adj=[[],[5],[6],[4],[],[6],[3,7],[]]
        self.assertNodeDisjoint(adj,3)
        self.assertGeneral(adj,2)

if __name__=='__main__':
    unittest.main()
